ua-snap/data-api
# Add filter-by-tag support to community endpoints

This PR adds filter-by-tag support to community endpoints. It uses the new `all_boundaries:all_communities` layer that was uploaded to gs.earthmaps.io on 1/8/2025, which was generated from the latest iteration of tagged community CSVs from the geospatial-vector-veracity repo's recently-merged https://github.com/ua-snap/geospatial-vector-veracity/pull/133.

This PR works by loading the tagged communities from `all_boundaries:all_communities` and filtering the results if tags are provided via the API's new `tags` HTTP GET parameter.

To test, run the API like this:

```
export FLASK_APP=application.py
export API_GS_BASE_URL=https://gs.earthmaps.io/geoserver/
export API_RAS_BASE_URL=https://zeus.snap.uaf.edu/rasdaman/
pipenv run flask run
```

Then reference the "tags" column of [these CSV files](https://github.com/ua-snap/geospatial-vector-veracity/tree/main/vector_data/point) to see which communities apply to which webapps (`ncr`, `eds`, `ardac`). Generally, they follow these rules:

- Arctic-EDS-exclusive communities (i.e., extra military sites) are tagged with just the `eds` tag.  EDS-exclusive communities are defined here https://github.com/ua-snap/geospatial-vector-veracity/blob/main/utilities/tag_point_locations.py#L28-L59
- Every community from every country's CSV, except for EDS-exclusive sites, are tagged with `ardac` because ARDAC Explorer allows exploration of pan-Arctic datasets.
- Communities from the Alaska, Yukon, and British Columbia CSVs are tagged with the `ncr` tag if, and only if, the lat/lon coordinate falls within the IEM extent polygon stored in the GVV repo. We were already performing this spatial check previously during shapefile generation, but this is now handled via tagging instead.

For example, "Nike Alaska Mike" is one of the locations that is tagged with just `eds`, meaning it should show up as a selectable place in just Arctic-EDS, not Northern Climate Reports or ARDAC Explorer.

Notice that "Nike Alaska Mike" shows up here:

http://localhost:5000/places/all
http://localhost:5000/places/all?tags=eds
http://localhost:5000/places/communities
http://localhost:5000/places/communities?tags=eds

But not here:

http://localhost:5000/places/all?tags=ncr
http://localhost:5000/places/communities?tags=ncr
http://localhost:5000/places/all?tags=ardac
http://localhost:5000/places/communities?tags=ardac

Likewise, "Nike Alaska Mike" shows up here:

http://localhost:5000/places/search/65/-147
http://localhost:5000/places/search/65/-147?tags=eds

But not here:

http://localhost:5000/places/search/65/-147?tags=ncr
http://localhost:5000/places/search/65/-147?tags=ardac

With this new feature in place, all we'll need to do is add the appropriate `?tags=...` parameter to the `/places/...` URLs of the corresponding webapps to make sure only communities with that webapp's tag show up as selectable places. Tickets for this have already been created for each affected webapp:

https://github.com/ua-snap/ardac-explorer/issues/183
https://github.com/ua-snap/northern-climate-reports/issues/688
https://github.com/ua-snap/arctic-eds/issues/448
